### PR TITLE
expression: implement vectorized evaluation for builtinLeastTimeSig

### PR DESCRIPTION
**Pull request by ylht**
### What problem does this PR solve? 
implement vectorized evaluation for `builtinLeastTimeSig`

### What is changed and how it works?
I change it from builtinGreatestTimeSig. I change the fisrt value from zerodatatime to maxdatetime and change the compare symbol from > to <, and deal with the invalid values. So it can support Least Time by vec, the speed is
```
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastTimeSig-VecBuiltinFunc-4                   412           2676687 ns/op          785814 B/op      43840 allocs/op
BenchmarkVectorizedBuiltinCompareFunc/builtinLeastTimeSig-NonVecBuiltinFunc-4                484           2454012 ns/op          700870 B/op      39465 allocs/op
```

### Check List 

Tests 

 - Unit test

### CODE REVIEW

**Comment by ylht** [comment]
> `builtinLeastTimeSig` and `builtinGreatestTimeSig` are not exactly the same. In `builtinLeastTimeSig`, it also check whether the function find invalid time.

I have committed a new version to deal with the invalid time.

File: expression/builtin_compare_vec.go [hunk 1]
```diff
@@ -707,11 +706,70 @@ func (b *builtinGreatestRealSig) vecEvalReal(input *chunk.Chunk, result *chunk.C
...
+			}
+			if argTime.Compare(dstTimes[i]) < 0 {
+				dstTimes[i] = argTime
+			}
+		}
+	}
+	result.ReserveString(n)
+	for i := 0; i < n; i++ {
+		if findInvalidTime[i] {
+			result.AppendString(invalidValue[0])
```
Comment (francis0407) [inline expression/builtin_compare_vec.go]: This implementation is not correct here.  The first string in `invalidValue` is not the one that `findInvalidTime` is true with the smallest `i`. For example, we have two vectors: `[valid, valid, invalidAAA], [valid, InvalidBBB, valid]`. findInvalidTime is:
`[false, true, true]`, but the invalidValue is: `[invalidAAA, invalidBBB]`.

File: expression/builtin_compare_vec.go [hunk 2]
```diff
@@ -707,11 +706,70 @@ func (b *builtinGreatestRealSig) vecEvalReal(input *chunk.Chunk, result *chunk.C
...
+	var argTime types.Time
+
+	var findInvalidTime []bool = make([]bool, n)
+	var invalidValue []string
+
+	for j := 0; j < len(b.args); j++ {
+		if err := b.args[j].VecEvalString(b.ctx, input, result); err != nil {
+			return err
+		}
+		for i := 0; i < n; i++ {
```
Comment (francis0407) [inline expression/builtin_compare_vec.go]: Use `result.MergeNulls(result, dst)` out of the loop.

Comment (francis0407) [reply expression/builtin_compare_vec.go]: Please address this comment.  @ylht 
You should handle the nulls like this:
```
dst.MergeNulls(result)
for i := 0; i < n; i++ {
    if dst.IsNull(i) {
        continue 
    }
}
```

File: expression/builtin_compare_vec_test.go [hunk 3]
```diff
@@ -134,6 +134,7 @@ var vecBuiltinCompareCases = map[string][]vecExprBenchCase{
 		{retEvalType: types.ETInt, childrenTypes: []types.EvalType{types.ETInt, types.ETInt, types.ETInt}},
 		{retEvalType: types.ETReal, childrenTypes: []types.EvalType{types.ETReal, types.ETReal, types.ETReal}},
 		{retEvalType: types.ETString, childrenTypes: []types.EvalType{types.ETString, types.ETString}},
+		{retEvalType: types.ETString, childrenTypes: []types.EvalType{types.ETDatetime, types.ETDatetime}},
```
Comment (francis0407) [inline expression/builtin_compare_vec_test.go]: Try use three children here.

**Comment by francis0407** [comment]
Could you please update the benchmark result? I think it could be better.

**Comment by ylht** [comment]
> Could you please update the benchmark result? I think it could be better.

I have updated it.

**Comment by qw4990** [comment]
> The performance is only changed a little. But I have no idea how to improve the implementation.
> 
> LGTM.

The two lines below do a large allocation each time when this function is called, so a way to optimize it is to make a pool for them like what we do for column buffers:
```
	var findInvalidTime []bool = make([]bool, n)
	var invalidValue []string = make([]string, n)
```
But for simplifying, we can optimize it later in another PR.

**Pull request merged**